core/services/ocr2/plugins/ocr2vrf/fixed_blockhash_provider.go

```diff
@@ -0,0 +1,109 @@
+package ocr2vrf
+
+import (
+	"context"
+	"math/big"
+	"time"
+
+	"github.com/pkg/errors"
+
+	"github.com/ethereum/go-ethereum/common"
+	"github.com/ethereum/go-ethereum/common/hexutil"
+	"github.com/ethereum/go-ethereum/rpc"
+	"github.com/smartcontractkit/ocr2vrf/types"
+
+	"github.com/smartcontractkit/chainlink/core/chains/evm/client"
+)
+
+// Implementation of Blockhashes interface that returns blockhashes with fixed-sized windows relative to current block height
+type FixedBlockhashProvider struct {
+	client client.Client
+	// end block = current head - waitBlocks
+	waitBlocks uint
+	// start block = current head - lookbackBlocks
+	lookbackBlocks uint
+	// number of blocks to query in a batch
+	batchSize uint
+}
+
+type Block struct {
+	Number    int64
+	Hash      common.Hash
+	Timestamp time.Time
+}
+
+var _ types.Blockhashes = (*FixedBlockhashProvider)(nil)
+
+func NewFixedBlockhashProvider(client client.Client, waitBlocks, lookbackBlocks, batchSize uint) *FixedBlockhashProvider {
+	return &FixedBlockhashProvider{
+		client,
+		waitBlocks,
+		lookbackBlocks,
+		batchSize,
+	}
+}
+
+func (b *FixedBlockhashProvider) OnchainVerifiableBlocks(
+	ctx context.Context,
+) (startHeight uint64, hashes []common.Hash, err error) {
+	height, err := b.CurrentHeight(ctx)
+	if err != nil {
+		return 0, nil, err
+	}
+	fromBlock := height - uint64(b.lookbackBlocks)
+	toBlock := height - uint64(b.waitBlocks)
+	reqs := []rpc.BatchElem{}
+
+	for i := fromBlock; i <= toBlock; i++ {
+		req := rpc.BatchElem{
+			Method: "eth_getBlockByNumber",
+			Args:   []interface{}{hexutil.EncodeBig(big.NewInt(int64(i))), true},
+			Result: &Block{},
+		}
+		reqs = append(reqs, req)
+	}
+
+	for i := 0; i < len(reqs); i += int(b.batchSize) {
+		j := i + int(b.batchSize)
+		if j > len(reqs) {
+			j = len(reqs)
+		}
+
+		err := b.client.BatchCallContext(ctx, reqs[i:j])
+
+		if err != nil {
+			return 0, nil, err
+		}
+	}
+
+	blockhashes := []common.Hash{}
+	for _, r := range reqs {
+		if r.Error != nil {
+			return 0, nil, r.Error
+		}
+		block, is := r.Result.(*Block)
+		if !is {
+			return 0, nil, errors.Errorf("expected result to be a %T, got %T", &Block{}, r.Result)
+		}
+		if block == nil {
+			return 0, nil, errors.New("invariant violation: got nil block")
+		}
+		if block.Hash == (common.Hash{}) {
+			return 0, nil, errors.New("missing block hash")
+		}
+		blockhashes = append(blockhashes, block.Hash)
+	}
+
+	return fromBlock, blockhashes, nil
+}
+
+func (b *FixedBlockhashProvider) CurrentHeight(ctx context.Context) (uint64, error) {
+	h, err := b.client.HeadByNumber(ctx, nil)
+	if err != nil {
+		return 0, err
```

Comment:
Lets wrap this error here: `errors.Wrap(err, "head by number")` to add more context to the error.